In [1]:
import  pandas as pd
import requests
import psycopg2
from sqlalchemy import create_engine, text
from io import StringIO

In [2]:
key = 'WJW4T8XOUIWC9AR9'
company = 'IBM'
url = f'https://www.alphavantage.co/query?function=TIME_SERIES_DAILY_ADJUSTED&symbol={company}&apikey={key}&outputsize=full'
r = requests.get(url)

In [3]:
json = r.json()
json.keys()


dict_keys(['Meta Data', 'Time Series (Daily)'])

In [4]:
df = pd.DataFrame(json['Time Series (Daily)'])
df

2023-02-08 2023-02-07 2023-02-06 2023-02-03 2023-02-02  \
1. open                  135.71     135.67     135.83     136.35     135.96   
2. high                  136.74      136.4     136.32     136.95     136.72   
3. low                   135.16     134.45     134.95     135.53     134.85   
4. close                 135.98     135.84     136.18     136.94     136.39   
5. adjusted close        135.98     135.84     136.18     136.94     136.39   
6. volume               4593748    3737553    4841300    3755720    6107793   
7. dividend amount       0.0000     0.0000     0.0000     0.0000     0.0000   
8. split coefficient        1.0        1.0        1.0        1.0        1.0   

                     2023-02-01 2023-01-31 2023-01-30 2023-01-27 2023-01-26  \
1. open                  134.49      135.5     134.32     134.44     137.53   
2. high                  135.79     135.65     136.11    135.488     138.27   
3. low                    132.8     133.76     133.98   133.7701     132.98   
4. close                 135.09     134.73      135.3     134.39     134.45   
5. adjusted close        135.09     134.73      135.3     134.39     134.45   
6. volume               5428898    7206448    5375712    8143146   17548483   
7. dividend amount       0.0000     0.0000     0.0000     0.0000     0.0000   
8. split coefficient        1.0        1.0        1.0        1.0        1.0   

                      ...        1999-11-12        1999-11-11  \
1. open               ...              96.0              97.5   
2. high               ...             96.25              97.5   
3. low                ...              93.0             94.87   
4. close              ...             95.87              95.0   
5. adjusted close     ...  52.8929457506281  52.4129534401759   
6. volume             ...           9668400           7147500   
7. dividend amount    ...            0.0000            0.0000   
8. split coefficient  ...               1.0               1.0   

                            1999-11-10       1999-11-09        1999-11-08  \
1. open                          94.62            95.06              90.5   
2. high                          97.06             95.5             93.94   
3. low                           94.12            92.12              90.5   
4. close                          97.0            93.62             93.94   
5. adjusted close     53.5163840389165  51.651586327045  51.8281352228435   
6. volume                      8680300          9551200           8522300   
7. dividend amount              0.0000           0.0000            0.1200   
8. split coefficient               1.0              1.0               1.0   

                            1999-11-05        1999-11-04        1999-11-03  \
1. open                          92.75             94.44             95.87   
2. high                          92.94             94.44             95.94   
3. low                           90.19              90.0              93.5   
4. close                         90.25             91.56             94.37   
5. adjusted close     49.7287816698025  50.4506066447326  51.9989487665292   
6. volume                     13737600          16697600          10369100   
7. dividend amount              0.0000            0.0000            0.0000   
8. split coefficient               1.0               1.0               1.0   

                            1999-11-02        1999-11-01  
1. open                          96.75              98.5  
2. high                          96.81             98.81  
3. low                           93.69             96.37  
4. close                         94.81             96.75  
5. adjusted close     52.2413937962767  53.3103559728907  
6. volume                     11105400           9551800  
7. dividend amount              0.0000            0.0000  
8. split coefficient               1.0               1.0  

[8 rows x 5856 columns]

In [5]:
time_col = 'timestamp'
df = df.transpose()
df=df.astype('float')


df=df.rename_axis(time_col).reset_index()
df[time_col]= pd.to_datetime(df[time_col])
df

timestamp  1. open  2. high  3. low  4. close  5. adjusted close  \
0    2023-02-08   135.71   136.74  135.16    135.98         135.980000   
1    2023-02-07   135.67   136.40  134.45    135.84         135.840000   
2    2023-02-06   135.83   136.32  134.95    136.18         136.180000   
3    2023-02-03   136.35   136.95  135.53    136.94         136.940000   
4    2023-02-02   135.96   136.72  134.85    136.39         136.390000   
...         ...      ...      ...     ...       ...                ...   
5851 1999-11-05    92.75    92.94   90.19     90.25          49.728782   
5852 1999-11-04    94.44    94.44   90.00     91.56          50.450607   
5853 1999-11-03    95.87    95.94   93.50     94.37          51.998949   
5854 1999-11-02    96.75    96.81   93.69     94.81          52.241394   
5855 1999-11-01    98.50    98.81   96.37     96.75          53.310356   

       6. volume  7. dividend amount  8. split coefficient  
0      4593748.0                 0.0                   1.0  
1      3737553.0                 0.0                   1.0  
2      4841300.0                 0.0                   1.0  
3      3755720.0                 0.0                   1.0  
4      6107793.0                 0.0                   1.0  
...          ...                 ...                   ...  
5851  13737600.0                 0.0                   1.0  
5852  16697600.0                 0.0                   1.0  
5853  10369100.0                 0.0                   1.0  
5854  11105400.0                 0.0                   1.0  
5855   9551800.0                 0.0                   1.0  

[5856 rows x 9 columns]

In [6]:
url_vm = "postgresql+psycopg2://postgres:a5TKQ7SSVGFuQIz5@primeds.protal.biz:30308/postgres"
engine = create_engine(url_vm)


Postgres

In [7]:
# table_name = 'stocks'
# df.head(0).to_sql(table_name, engine, if_exists='replace',index=False)
# conn = engine.raw_connection()
# cur = conn.cursor()
# output = StringIO()
# df.to_csv(output, sep='\t', header=False, index=False)
# output.seek(0)
# contents = output.getvalue()
# cur.copy_from(output, table_name, null="") # null values become ''
# conn.commit()
# cur.close()
# conn.close()

Timescaledb

In [8]:
table_name = 'timetest1'
schema = 'staging'

query = f"SELECT create_hypertable('{schema}.{table_name}', 'timestamp',migrate_data => true);"

df.head(0).to_sql(table_name, engine, if_exists='replace',index=False, schema=schema)
conn = engine.raw_connection()
cur = conn.cursor()
output = StringIO()
df.to_csv(output, sep='\t', header=False, index=False)
output.seek(0)
contents = output.getvalue()
cur.copy_expert(f'COPY {schema}.{table_name} FROM STDIN', output)
cur.execute(query)

conn.commit()
cur.close()
conn.close()

3.18 s ± 140 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


### Load function time

CPU times: total: 0 ns
Wall time: 209 ms

### Whole cell time

3.18 s ± 140 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)